src/Compilers/Core/Portable/Syntax/ChildSyntaxList.cs

```diff
@@ -83,17 +83,46 @@ private static int Occupancy(GreenNode green)
             return green.IsList ? green.SlotCount : 1;
         }
 
+        internal struct SlotData(int slotIndex, int precedingOccupantSlotCount, int positionAtSlotIndex)
+        {
+            /// <summary>
+            /// The green node slot index at which to start the search
+            /// </summary>
+            public int SlotIndex { get; set; } = slotIndex;
+
+            /// <summary>
+            /// Indicates the total number of occupants in preceding slots
+            /// </summary>
+            public int PrecedingOccupantSlotCount { get; set; } = precedingOccupantSlotCount;
+
+            /// <summary>
+            /// Indicates the node start position plus any prior slot full widths
+            /// </summary>
+            public int PositionAtSlotIndex { get; set; } = positionAtSlotIndex;
+        }
+
+        internal static SyntaxNodeOrToken ItemInternal(SyntaxNode node, int index)
+        {
+            var slotData = new SlotData(slotIndex: 0, precedingOccupantSlotCount: 0, positionAtSlotIndex: node.Position);
+
+            return ItemInternal(node, index, ref slotData);
+        }
+
         /// <summary>
         /// internal indexer that does not verify index.
         /// Used when caller has already ensured that index is within bounds.
         /// </summary>
-        internal static SyntaxNodeOrToken ItemInternal(SyntaxNode node, int index)
+        internal static SyntaxNodeOrToken ItemInternal(SyntaxNode node, int index, ref SlotData slotData)
         {
             GreenNode? greenChild;
             var green = node.Green;
-            var idx = index;
-            var slotIndex = 0;
-            var position = node.Position;
+
+            // slotData may contain information that allows us to start the loop below using data
+            // calculated during a previous call. As index represents the offset into all children of
+            // node, idx represents the offset requested relative to the given slot index.
+            var idx = index - slotData.PrecedingOccupantSlotCount;
```

Comment:
>var idx = index - slotData.PrecedingOccupantSlotCount; [](http://example.com/codeflow?start=12&length=54)

Are we expecting `index >= slotData.PrecedingOccupantSlotCount`? If so, consider asserting that.

Reply:
Done!